pages/signin.js

```diff
@@ -16,36 +16,41 @@ import SignInOrJoinFree from '../components/SignInOrJoinFree';
 import { withUser } from '../components/UserProvider';
 
 class SigninPage extends React.Component {
-  static getInitialProps({ query: { token, next, form } }) {
+  static getInitialProps({ query: { token, next, form, tfa } }) {
     // Decode next URL if URI encoded
     if (next && next.startsWith('%2F')) {
       next = decodeURIComponent(next);
     }
 
     next = next && isValidRelativeUrl(next) ? next : null;
-    return { token, next, form: form || 'signin' };
+    return { token, next, form: form || 'signin', tfa };
   }
 
   static propTypes = {
     form: PropTypes.oneOf(['signin', 'create-account']).isRequired,
     token: PropTypes.string,
     next: PropTypes.string,
+    tfa: PropTypes.string,
     login: PropTypes.func,
     errorLoggedInUser: PropTypes.string,
     LoggedInUser: PropTypes.object,
     loadingLoggedInUser: PropTypes.bool,
+    LoggedInUserTwoFactorAuth: PropTypes.bool,
   };
 
   state = { error: null, success: null };
 
   async componentDidMount() {
-    if (this.props.token) {
-      let user = await this.props.login(this.props.token);
+    if (this.props.token && !this.props.tfa) {
+      await this.props.login(this.props.token);
```

Comment:
Maybe this is where you should do a try/catch, wait for the exception and call ``await Router.pushRoute('signin', { token, tfa: true });` when needed. 

In a second step, maybe you'll realize that pushing the route is not really needed and you can just use the state of the component instead.

Reply:
That sounds like a good idea, thank you. A lot of frustration yesterday came from figuring out how to properly break out of this componentDidMount loop.